### PR TITLE
Answer: 5

### PR DESCRIPTION
**Pull request by dmmishchenko**
## Checklist for challenge submission

- [x] Start your PR message with Answer:${challenge_number}

## Link to challenge:
[🟢 Crud application](https://angular-challenges.vercel.app/challenges/angular/5-crud/)

### CODE REVIEW

File: apps/angular/crud/src/app/data-access/todos.store.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+      .get<Todo[]>('https://jsonplaceholder.typicode.com/todos')
+      .subscribe({
+        next: (todos) => {
+          this.patchState({ todos, loading: false });
+        },
+        error: (error) => {
+          this.patchState({ loading: false, error: JSON.stringify(error) });
+        },
+      });
+  }
```
Comment (tomalaforge) [inline apps/angular/crud/src/app/data-access/todos.store.ts:39]: if you want to use the component store, don't use subcribe but wrap your code inside an effect

File: apps/angular/crud/src/app/data-access/todos.store.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+      todos: [],
+      loading: false,
+      error: null,
+      disabledTodosIds: [],
+      errorTodosIds: [],
+    });
+  }
+
+  load() {
+    this.patchState({ loading: true, error: null });
```
Comment (tomalaforge) [inline apps/angular/crud/src/app/data-access/todos.store.ts:28]: you should create a property that have both loading and error state. Less error prone

`callstate = Loading | {err: stirng}`

File: apps/angular/crud/src/app/data-access/todos.store.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+              errorTodosIds: [...state.errorTodosIds, id],
+              disabledTodosIds: state.disabledTodosIds.filter(
+                (id) => id !== id
+              ),
+            };
+          });
+        },
+      });
+  }
+}
```
Comment (tomalaforge) [inline apps/angular/crud/src/app/data-access/todos.store.ts:125]: use `effect` for all your methods

File: apps/angular/crud/src/app/components/item.component.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      <button
+        aria-label="delete button"
+        [disabled]="disabledTodosIds?.includes(item.id)"
+        (click)="delete(item)">
+        Delete
+      </button>
+      <ng-container *ngIf="errorTodosIds?.includes(item.id)">
+        Error occuried
+      </ng-container>
+    </div>
```
Comment (tomalaforge) [inline apps/angular/crud/src/app/components/item.component.ts:35]: next step: loading should be handle per item as well.

File: apps/angular/crud/src/app/app.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  it('should call todosStore.upload on handleUpdate', () => {
+    component.handleUpdate({
+      id: 1,
+      userId: 1,
+      title: 'test',
+      completed: false,
+    });
+    expect(mockMoviesStore.update).toHaveBeenCalled();
+  });
+});
```
Comment (tomalaforge) [inline apps/angular/crud/src/app/app.component.spec.ts:47]: My personal opinion: This is great to have test, but tests using implementation details are very hard to maintain. You should either test the component as a whole, or the each method on the service.

File: apps/angular/crud/src/app/data-access/todos.store.ts (hunk 1 above)
Comment (dmmishchenko) [reply apps/angular/crud/src/app/data-access/todos.store.ts]: changed to effects

File: apps/angular/crud/src/app/data-access/todos.store.ts (hunk 2 above)
Comment (dmmishchenko) [reply apps/angular/crud/src/app/data-access/todos.store.ts]: changed to one property use

File: apps/angular/crud/src/app/data-access/todos.store.ts (hunk 3 above)
Comment (dmmishchenko) [reply apps/angular/crud/src/app/data-access/todos.store.ts:145]: done!

File: apps/angular/crud/src/app/components/item.component.ts (hunk 4 above)
Comment (dmmishchenko) [reply apps/angular/crud/src/app/components/item.component.ts:35]: can you please provide a little more? You suggest create some inner store for each item?

**Comment by tomalaforge** [comment]
That's perfect for the store. 

For the local loading, you can create a local service thanks to the injection context, meaning using the proverder property inside the component devcorator

**Comment by dmmishchenko** [comment]
@tomalaforge how about this? 
I created local service and work with item update/delete logic inside it. 
[Link to changes](https://github.com/tomalaforge/angular-challenges/pull/262/commits/0736609298553bdece4bd637811101c00720400f)

**Comment by tomalaforge** [comment]
yes that's perfect 👍

**Pull request closed without merging**